pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+				con.mu.RUnlock()
+
+				if vHostNonceSent != "" && vHostNonceSent != discReq.ResponseNonce {
+					adsLog.Debugf("ADS:RDS: Expired nonce received %s %s (%v), sent %s, received %s",
+						peerAddr, con.ConID, con.modelNode, vHostNonceSent, discReq.ResponseNonce)
+					vhdsExpiredNonce.Inc()
+					continue
+				}
+
+				//BAVERY_TODO: Add ErrorDetail
+				//BAVERY_QUESTION: Do we support namespace in the path? Probably not.
+
+				for _, subscribeVhost := range subscribeHosts {
+					routeName, host, err := splitVHost(subscribeVhost)
+					if err != nil {
+						adsLog.Errorf("Request received with invalid route config name: %+v.", err.Error())
+					} else {
+						//use wildcard namespace to select the service from any available namespace. We likely can't get the actual namespace from Envoy.
+						sscope, err := subscribeToVHost(con.modelNode.SidecarScope, s.globalPushContext(), routeName, "*/"+host, "*")
+						if err != nil {
+							fmt.Printf("Failed to subscribe to vhost: %s", err.Error())
+							continue
+						}
+						con.modelNode.SidecarScope = sscope
+					}
+				}
+
+				err := s.pushDeltaVirtualHost(con, s.globalPushContext(), unsubscribeHosts)
+				if err != nil {
+					return err
+				}
+
+			case EndpointType:
+				return status.Errorf(codes.Unimplemented, "not implemented")
+			default:
+				adsLog.Warnf("ADS: Unknown watched resources %s", discReq.String())
+			}
+
+			if !con.added {
+				con.added = true
+				s.addCon(con.ConID, con)
+				defer s.removeCon(con.ConID, con)
+			}
+		case pushEv := <-con.pushChannel:
+			// It is called when config changes.
+			// This is not optimized yet - we should detect what changed based on event and only
+			// push resources that need to be pushed.
+
+			// TODO: possible race condition: if a config change happens while the envoy
+			// was getting the initial config, between LDS and RDS, the push will miss the
+			// monitored 'routes'. Same for CDS/EDS interval.
+			// It is very tricky to handle due to the protocol - but the periodic push recovers
+			// from it.
+
+			err := s.pushConnection(con, pushEv)
+			if err != nil {
+				return nil
+			}
+
+		}
+	}
+}
+
+/*func mergeSubscribeUnsubsribe(resource []string, subscribe []string, unsubscribe []string) []string {
+	for i := 0; i < len(resource); i++ {
+		for j := 0; j < len(subscribe); j++ {
+			for k := 0; k < len(unsubscribe); k++ {
+
+			}
+		}
+	}
+}*/
+
+// Send with timeout
+func (conn *XdsConnection) sendDelta(res *xdsapi.DeltaDiscoveryResponse) error {
+	done := make(chan error)
+	t := time.NewTimer(SendTimeout)
+	go func() {
+		err := conn.deltaStream.Send(res)
+		done <- err
+		conn.mu.Lock()
+		if res.Nonce != "" {
+			conn.RouteNonceSent = res.Nonce
+		}
+		conn.mu.Unlock()
+	}()
+	select {
+	case <-t.C:
+		// TODO: wait for ACK
+		adsLog.Infof("Timeout writing %s", conn.ConID)
+		xdsResponseWriteTimeouts.Add(1)
+		return errors.New("timeout sending")
+	case err := <-done:
+		t.Stop()
+		return err
+	}
+}
+
```

Comment:
Comments ? Seems strange to return the *SidecarScope - I assume you want to add the dynamic resources to the scope ? That would just modify the parameter.

A better pattern may be to add this as a method to SidecarScope. 

Please make sure locks are in place if you modify the resources.

Making copies of the sidecar is not a good idea.